### ISSUE TITLE
Cassandra docker image exits

### ISSUE THREAD

**Issue by Daerdemandt** [yasp-dota/yasp#1121]
Attempting to run as readme suggests (using Docker) fails: Cassandra image exits almost instantly so attempt to init it (`sudo  docker exec -i cassandra cqlsh < sql/cassandra.cql`) generates an error:

    Error response from daemon: Container ... is not running

Here's what `docker ps -a` says after launching Cassandra:

    CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                     PORTS               NAMES
    7a481334226d        cassandra:3         "/docker-entrypoint.s"   13 seconds ago      Exited (3) 1 seconds ago                       cassandra
    c107cff30103        redis:3             "docker-entrypoint.sh"   18 seconds ago      Up 16 seconds                                  redis
    bbf4d10a436e        postgres:9.5        "/docker-entrypoint.s"   22 seconds ago      Up 20 seconds                                  postgres
    4d87dc318410        yasp/yasp:latest    "npm start"              32 seconds ago      Up 29 seconds                                  yasp

If I continue the process ignoring Cassandra fail, I get the web interface running but requesting a replay to be parsed fails. *This* is my main concern, but I suppose it's exactly because of a failure with Cassandra - as I understand it, Cassandra is not optional [anymore](https://github.com/yasp-dota/yasp/commit/e37323955bb8bf178f6fdb96072cd40eddb788ed), hence the issue. 

`docker info` says:

    Containers: 6
     Running: 3
     Paused: 0
     Stopped: 3
    Images: 202
    Server Version: 1.11.2
    Storage Driver: devicemapper
     Pool Name: docker-8:8-5242884-pool
     Pool Blocksize: 65.54 kB
     Base Device Size: 107.4 GB
     Backing Filesystem: ext4
     Data file: /dev/loop0
     Metadata file: /dev/loop1
     Data Space Used: 10.3 GB
     Data Space Total: 107.4 GB
     Data Space Available: 97.07 GB
     Metadata Space Used: 19.21 MB
     Metadata Space Total: 2.147 GB
     Metadata Space Available: 2.128 GB
     Udev Sync Supported: true
     Deferred Removal Enabled: false
     Deferred Deletion Enabled: false
     Deferred Deleted Device Count: 0
     Data loop file: /home/docker/docker-data/devicemapper/devicemapper/data
     WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
     Metadata loop file: /home/docker/docker-data/devicemapper/devicemapper/metadata
     Library Version: 1.02.126 (2016-06-03)
    Logging Driver: json-file
    Cgroup Driver: cgroupfs
    Plugins: 
     Volume: local
     Network: host bridge null
    Kernel Version: 4.6.2-1-ARCH
    Operating System: Arch Linux
    OSType: linux
    Architecture: x86_64
    CPUs: 2
    Total Memory: 3.863 GiB
    ID: RIZS:JGNG:OQJU:HV4B:O3Y6:XOXN:RMIQ:6JYX:ODR5:GOSS:4XWO:GVDK
    Docker Root Dir: /home/docker/docker-data
    Debug mode (client): false
    Debug mode (server): false
    Registry: https://index.docker.io/v1/

If it's not me and there really *is* a problem, it should have been present for more than 3 weeks. I've checked other issues and haven't found anything about this one. Does anybody else experience the problem?

**Comment by howardchung** [comment]
What do the logs say?

`sudo docker logs cassandra`

**Comment by mintwzy** [comment]
I get `hostname: Name or service not known` with `docker logs cassandra` on OS X EI Capitan with 16GB RAM.

**Comment by howardchung** [comment]
And this is after starting the container?

**Comment by mintwzy** [comment]
Yes, after `docker start cassandra`

**Comment by howardchung** [comment]
Did you start with:
`sudo docker run -d --name cassandra --net=host cassandra:3`

Otherwise, your image may not be named "cassandra" and you would have to use the container ID

**Comment by mintwzy** [comment]
I start it with `docker run -d --name cassandra --net=host cassandra:3`, then got 
```
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                      PORTS               NAMES
1c22594d6e77        cassandra:3         "/docker-entrypoint.s"   3 seconds ago       Exited (1) 2 seconds ago                        cassandra
```
When I use `docker logs cassandra`, I got `hostname: Name or service not known`. 

There is no such a problem several weeks ago if my memory serves me well.

**Comment by howardchung** [comment]
Try setting the version specifically, 3.5

Looks like Cassandra recently updated to 3.7 and the image may have a different start command.

https://hub.docker.com/_/cassandra/

**Comment by mintwzy** [comment]
Still the same problem with `docker run -d --name cassandra --net=host cassandra:3.5`.

**Comment by howardchung** [comment]
Are you running directly on OS X with no linux VM?

**Comment by mintwzy** [comment]
Directly on OS X without linux VM, could this be a problem?

**Comment by howardchung** [comment]
Yeah, give it a try on a linux VM and see if you have the same problem.  I do not have OS X so I can't repro

**Comment by mintwzy** [comment]
Sure, let me try with `Ubuntu 16.04`.

**Comment by mintwzy** [comment]
Not a problem on `Ubuntu 16.04` with `sudo docker run -d --name cassandra2 --net=host cassandra:3`

**Comment by mintwzy** [comment]
This is a solution for me, from docker-library/cassandra#32

**Comment by Daerdemandt** [comment]
@howardchung 
> What do the logs say?

    Unknown listen_address '::1 127.0.0.1'

The same thing for either of `cassandra:3`, `:3.5` and `:3.7`

Both `::1` and `127.0.0.1` are present in my `/etc/hosts` if that is relevant:

    127.0.0.1	localhost.localdomain	localhost 
    ::1		localhost.localdomain	localhost

Full log dump:

    INFO  11:52:08 Configuration location: file:/etc/cassandra/cassandra.yaml
    INFO  11:52:08 Node configuration:[allocate_tokens_for_keyspace=null; authenticator=AllowAllAuthenticator; authorizer=AllowAllAuthorizer; auto_bootstrap=true; auto_snapshot=true; batch_size_fail_threshold_in_kb=50; batch_size_warn_threshold_in_kb=5; batchlog_replay_throttle_in_kb=1024; broadcast_address=::1 127.0.0.1; broadcast_rpc_address=::1 127.0.0.1; buffer_pool_use_heap_if_exhausted=true; cas_contention_timeout_in_ms=1000; client_encryption_options=<REDACTED>; cluster_name=Test Cluster; column_index_cache_size_in_kb=2; column_index_size_in_kb=64; commit_failure_policy=stop; commitlog_compression=null; commitlog_directory=/var/lib/cassandra/commitlog; commitlog_max_compression_buffers_in_pool=3; commitlog_periodic_queue_size=-1; commitlog_segment_size_in_mb=32; commitlog_sync=periodic; commitlog_sync_batch_window_in_ms=null; commitlog_sync_period_in_ms=10000; commitlog_total_space_in_mb=null; compaction_large_partition_warning_threshold_mb=100; compaction_throughput_mb_per_sec=16; concurrent_compactors=null; concurrent_counter_writes=32; concurrent_materialized_view_writes=32; concurrent_reads=32; concurrent_replicates=null; concurrent_writes=32; counter_cache_keys_to_save=2147483647; counter_cache_save_period=7200; counter_cache_size_in_mb=null; counter_write_request_timeout_in_ms=5000; credentials_cache_max_entries=1000; credentials_update_interval_in_ms=-1; credentials_validity_in_ms=2000; cross_node_timeout=false; data_file_directories=[Ljava.lang.String;@7219ec67; disk_access_mode=auto; disk_failure_policy=stop; disk_optimization_estimate_percentile=0.95; disk_optimization_page_cross_chance=0.1; disk_optimization_strategy=ssd; dynamic_snitch=true; dynamic_snitch_badness_threshold=0.1; dynamic_snitch_reset_interval_in_ms=600000; dynamic_snitch_update_interval_in_ms=100; enable_scripted_user_defined_functions=false; enable_user_defined_functions=false; enable_user_defined_functions_threads=true; encryption_options=null; endpoint_snitch=SimpleSnitch; file_cache_size_in_mb=null; gc_warn_threshold_in_ms=1000; hinted_handoff_disabled_datacenters=[]; hinted_handoff_enabled=true; hinted_handoff_throttle_in_kb=1024; hints_compression=null; hints_directory=null; hints_flush_period_in_ms=10000; incremental_backups=false; index_interval=null; index_summary_capacity_in_mb=null; index_summary_resize_interval_in_minutes=60; initial_token=null; inter_dc_stream_throughput_outbound_megabits_per_sec=200; inter_dc_tcp_nodelay=false; internode_authenticator=null; internode_compression=dc; internode_recv_buff_size_in_bytes=null; internode_send_buff_size_in_bytes=null; key_cache_keys_to_save=2147483647; key_cache_save_period=14400; key_cache_size_in_mb=null; listen_address=::1 127.0.0.1; listen_interface=null; listen_interface_prefer_ipv6=false; listen_on_broadcast_address=false; max_hint_window_in_ms=10800000; max_hints_delivery_threads=2; max_hints_file_size_in_mb=128; max_mutation_size_in_kb=null; max_streaming_retries=3; max_value_size_in_mb=256; memtable_allocation_type=heap_buffers; memtable_cleanup_threshold=null; memtable_flush_writers=1; memtable_heap_space_in_mb=null; memtable_offheap_space_in_mb=null; native_transport_max_concurrent_connections=-1; native_transport_max_concurrent_connections_per_ip=-1; native_transport_max_frame_size_in_mb=256; native_transport_max_threads=128; native_transport_port=9042; native_transport_port_ssl=null; num_tokens=256; otc_coalescing_strategy=TIMEHORIZON; otc_coalescing_window_us=200; partitioner=org.apache.cassandra.dht.Murmur3Partitioner; permissions_cache_max_entries=1000; permissions_update_interval_in_ms=-1; permissions_validity_in_ms=2000; phi_convict_threshold=8.0; prepared_statements_cache_size_mb=null; range_request_timeout_in_ms=10000; read_request_timeout_in_ms=5000; request_scheduler=org.apache.cassandra.scheduler.NoScheduler; request_scheduler_id=null; request_scheduler_options=null; request_timeout_in_ms=10000; role_manager=CassandraRoleManager; roles_cache_max_entries=1000; roles_update_interval_in_ms=-1; roles_validity_in_ms=2000; row_cache_class_name=org.apache.cassandra.cache.OHCProvider; row_cache_keys_to_save=2147483647; row_cache_save_period=0; row_cache_size_in_mb=0; rpc_address=0.0.0.0; rpc_interface=null; rpc_interface_prefer_ipv6=false; rpc_keepalive=true; rpc_listen_backlog=50; rpc_max_threads=2147483647; rpc_min_threads=16; rpc_port=9160; rpc_recv_buff_size_in_bytes=null; rpc_send_buff_size_in_bytes=null; rpc_server_type=sync; saved_caches_directory=/var/lib/cassandra/saved_caches; seed_provider=org.apache.cassandra.locator.SimpleSeedProvider{seeds=::1 127.0.0.1}; server_encryption_options=<REDACTED>; snapshot_before_compaction=false; ssl_storage_port=7001; sstable_preemptive_open_interval_in_mb=50; start_native_transport=true; start_rpc=false; storage_port=7000; stream_throughput_outbound_megabits_per_sec=200; streaming_socket_timeout_in_ms=86400000; thrift_framed_transport_size_in_mb=15; thrift_max_message_length_in_mb=16; thrift_prepared_statements_cache_size_mb=null; tombstone_failure_threshold=100000; tombstone_warn_threshold=1000; tracetype_query_ttl=86400; tracetype_repair_ttl=604800; transparent_data_encryption_options=org.apache.cassandra.config.TransparentDataEncryptionOptions@45018215; trickle_fsync=false; trickle_fsync_interval_in_kb=10240; truncate_request_timeout_in_ms=60000; unlogged_batch_across_partitions_warn_threshold=10; user_defined_function_fail_timeout=1500; user_defined_function_warn_timeout=500; user_function_timeout_policy=die; windows_timer_interval=1; write_request_timeout_in_ms=2000]
    INFO  11:52:08 DiskAccessMode 'auto' determined to be mmap, indexAccessMode is mmap
    INFO  11:52:08 Global memtable on-heap threshold is enabled at 251MB
    INFO  11:52:08 Global memtable off-heap threshold is enabled at 251MB
    Exception (org.apache.cassandra.exceptions.ConfigurationException) encountered during startup: Unknown listen_address '::1 127.0.0.1'
    Unknown listen_address '::1 127.0.0.1'
    ERROR 11:52:08 Exception encountered during startup: Unknown listen_address '::1 127.0.0.1'

**Comment by howardchung** [comment]
Not sure, you could try:
* Debian-based OS
* Find a IP address and set CASSANDRA_LISTEN_ADDRESS as described in the linked issue

**Comment by howardchung** [comment]
any update on this?

**Comment by Daerdemandt** [comment]
I've used `docker run -d --name cassandra --net=host -e CASSANDRA_LISTEN_ADDRESS='127.0.0.1' cassandra:3` - that launches the container that doesn't exit immediately and subsequent initing command works.

There's also [another](http://stackoverflow.com/questions/38460998/cassandra-container-with-net-host-exits-immediately/38773505#38773505) option - removing `::1` line from `hosts` file works too.

I can not say that everything works fine - I get the webpage but requesting to parse a match still fails - but I guess it's because some other (yet unknown) reasons. If I look at Cassandra [logs](http://pastebin.com/aSfXY9J9), there's nothing that looks like error messages so I guess this issue is resolved.

I've tried on another machine with arch linux and problem with Cassandra was present there too and solutions worked there too. If your `hosts` file has `::1` line then I guess it's something distro-specific. If it does not, then maybe that's what causes different behaviour.

Also, speaking of low RAM: this indeed causes Cassandra container to exit right away, albeit with a different error message: Cassandra tries to allocate 4Gb at the start and if it fails, it fails. If someone is having similar problem, simply adding enough swap space works. I tried on a laptop with 4Gb RAM - it was failing without swap, but worked fine with one.

There's also a thing that docker-compose does not use Cassandra so it's not working now.
I'm going to get my hands on a machine with Debian-based OS tomorrow. If the project runs there, I'll be able to PR an update for docker-compose.
